src/core/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -276,16 +402,138 @@ function propVersion(doc: SavedObjectUnsanitizedDoc | ActiveMigrations, prop: st
+      attributes: {
+        targetNamespace: namespace,
+        targetType: type,
+        targetId: id,
+      },
+    };
+    additionalDocs.push(legacyUrlAlias);
+  }
+  return {
+    transformedDoc: { ...otherAttrs, id, originId, namespaces: [namespace] },
+    additionalDocs,
+  };
+}
+
+/**
+ * Returns all applicable conversion transforms for a given object type.
+ */
+function getConversionTransforms(type: SavedObjectsType): Transform[] {
+  const { convertToMultiNamespaceTypeVersion } = type;
+  if (!convertToMultiNamespaceTypeVersion) {
+    return [];
+  }
+  return [
+    {
+      version: convertToMultiNamespaceTypeVersion,
+      transform: convertType,
+      transformType: 'convert',
+    },
+  ];
+}
+
+/**
+ * Returns all applicable reference transforms for all object types.
+ */
+function getReferenceTransforms(typeRegistry: ISavedObjectTypeRegistry): Transform[] {
+  const transformMap = typeRegistry
+    .getAllTypes()
+    .filter((type) => type.convertToMultiNamespaceTypeVersion)
+    .reduce((acc, { convertToMultiNamespaceTypeVersion: key, name }) => {
+      const val = acc.get(key!) ?? new Set();
+      return acc.set(key!, val.add(name));
+    }, new Map<string, Set<string>>());
+
+  return Array.from(transformMap, ([key, val]) => ({
+    version: key,
+    transform: (doc) => {
+      const { namespace, references } = doc;
+      if (namespace && references?.length) {
+        return {
+          transformedDoc: {
+            ...doc,
+            references: references.map(({ type, id, ...attrs }) => ({
+              ...attrs,
+              type,
+              id: val.has(type) ? uuidv5(`${namespace}:${type}:${id}`, uuidv5.DNS) : id,
+            })),
+          },
+          additionalDocs: [],
+        };
+      }
+      return { transformedDoc: doc, additionalDocs: [] };
+    },
+    transformType: 'reference',
+  }));
+}
+
+/**
+ * Transforms are sorted in ascending order by version. One version may contain multiple transforms; 'reference' transforms always run
+ * first, 'convert' transforms always run second, and 'migrate' transforms always run last. This is because:
+ *  1. 'convert' transforms get rid of the `namespace` field, which must be present for 'reference' transforms to function correctly.
+ *  2. 'migrate' transforms are defined by the consumer, and may change the object type or migrationVersion which resets the migration loop
+ *     and could cause any remaining transforms for this version to be skipped.
+ */
+function transformComparator(a: Transform, b: Transform) {
+  const semver = Semver.compare(a.version, b.version);
+  if (semver !== 0) {
+    return semver;
+  } else if (a.transformType !== b.transformType) {
+    if (a.transformType === 'migrate') {
+      return 1;
+    } else if (b.transformType === 'migrate') {
+      return -1;
+    } else if (a.transformType === 'convert') {
+      return 1;
+    } else if (b.transformType === 'convert') {
+      return -1;
```

Comment:
Well the full snippet is

```ts
  const semver = Semver.compare(a.version, b.version);
  if (semver !== 0) {
    return semver;
  } else if (a.transformType !== b.transformType) {
    if (a.transformType === 'migrate') {
      return 1;
    } else if (b.transformType === 'migrate') {
      return -1;
    } else if (a.transformType === 'convert') {
      return 1;
    } else if (b.transformType === 'convert') {
      return -1;
    }
  }
  return 0;
```

So if `a.transformType === 'migrate' && b.transformType === 'migrate'` we would return `0`.

I know this comparator function is ugly as sin but I couldn't see a way to make it more elegant while still being readable. We want to make sure transforms are sorted so that "reference" transforms run first, "convert" transforms run second, and "migrate" transforms run last.

Reply:
Oh right, I missed the initial `if (a.transformType !== b.transformType)`, nevermind then